How does SingleCellFusion work?
================================
SingleCellFusion is built around the idea that for a cell profiled by a given omics technique (RNA-sequencing,
snATAC-sequencing, snmC-sequencing) there are unobserved features of that cell that if sampled would
provide a fuller picture of that cell's identity. For example, if a cell underwent RNA-sequencing we know
what genes are expressed but we don't know the patterns of DNA methylation in that same cell. The methylation
status of DNA in that cell is unobserved, limiting our ability to fully understand the identity of that cell.

In an ideal world we would obtain the transcriptome, methylome, and chromatin accessibility of a single
cell at once, but as the technologies for this type of experiment develop SingleCellFusion can provide a
computational equivalent. SingleCellFusion uses known relationships between different types of multiomics
data to impute unobserved data, enabling the multimodal analysis of a cell's identity.

The core of SingleCellFusion is the generation of a nearest neighbors graph between different data sets.
This graph is generated by finding nearest neighbors using the correlation of counts at highly variable
features. For example, DNA methylation is known to be negatively correlated with gene expression. If a
snmC-seq profiled cell has low methylation at a number of highly variable genes, and a snRNA-seq profiled
cell has high gene expression at those same genes, we can assume that those two cells likely belong to the
same cell type. We use this nearest neighbors graph to generate imputed counts by averaging among a cell's
neighbors in the opposite modality. The actions of SingleCellFusion depend on the type of nearest neighbor
graph specified, and are described below.

Direct mutual nearest neighbors
-------------------------------
This is the most conservative method for generating imputed counts, only cells that make direct mutual
nearest neighbors will receive imputed data. This method typically leads to good integrations but can
result in the loss of large fractions of cells from the analysis if mutual neighbors were not found for them.


Mutual nearest neighbors with rescue
-------------------------------------
This is a more lenient method for generating imputed counts, and all cells will receive imputed data. This
method will enable all cells to be analyzed, and is our recommended approach.

k-nearest neighbors
-------------------
This is the most relaxed method for generating imputed counts, as all cells will make neighbors in the
opposite data set.

